vim-patch:8.1.{199,200}

janlazo opened the pull request:
**vim-patch:8.1.0199: spellbadword() does not check for caps error**
Problem:    spellbadword() does not check for caps error. (Dominique Pelle)
Solution:   Adjust capcol when advancing.
vim/vim@66ab916

**vim-patch:8.1.0200: spellbadword() not tested**
Problem:    spellbadword() not tested.
Solution:   Add a test. (Dominique Pelle, closes vim/vim#3235)
vim/vim@872e451

janlazo committed:
vim-patch:8.1.0199: spellbadword() does not check for caps error

Problem:    spellbadword() does not check for caps error. (Dominique Pelle)
Solution:   Adjust capcol when advancing.
https://github.com/vim/vim/commit/66ab916935585391b2efaa8e39075e1ef94717b1

janlazo committed:
vim-patch:8.1.0200: spellbadword() not tested

Problem:    spellbadword() not tested.
Solution:   Add a test. (Dominique Pelle, closes vim/vim#3235)
https://github.com/vim/vim/commit/872e451e8c326d5dd3062ef621fcbf0a4c5bef78

justinmk commented:
```
[OLDTEST] Running test_startup
=================================================================
==29606==ERROR: AddressSanitizer: global-buffer-overflow on address 0x000001e1b221 at pc 0x0000004f22e5 bp 0x7ffc6fc3e830 sp 0x7ffc6fc3dfe0
READ of size 6 at 0x000001e1b221 thread T0
    #0 0x4f22e4 in __asan_memcpy (/home/travis/build/neovim/neovim/build/bin/nvim+0x4f22e4)
    #1 0x10f0573 in xmemdupz /home/travis/build/neovim/neovim/build/../src/nvim/memory.c:196:10
    #2 0xb0a47b in tv_list_append_string /home/travis/build/neovim/neovim/build/../src/nvim/eval/typval.c:553:28
    #3 0xa0b8f4 in f_spellbadword /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:15681:3
    #4 0x8e57c7 in call_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6406:11
    #5 0x8fa863 in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6148:11
    #6 0x987ed1 in eval7 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:4246:15
    #7 0x98446b in eval6 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3943:7
    #8 0x981d77 in eval5 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3794:7
    #9 0x97d07f in eval4 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3531:7
    #10 0x97c55c in eval3 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3449:7
    #11 0x97ba2c in eval2 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3380:7
    #12 0x8ddbe2 in eval1 /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:3307:7
    #13 0x8fa23d in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6121:9
    #14 0x8f33ee in ex_call /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:2743:9
    #15 0xc616d0 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2242:5
    #16 0xc3fe67 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:593:20
    #17 0x91b43c in call_user_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:21608:3
    #18 0x8e50eb in call_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6391:11
    #19 0x8fa863 in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6148:11
    #20 0x8f33ee in ex_call /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:2743:9
    #21 0xc616d0 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2242:5
    #22 0xc3fe67 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:593:20
    #23 0x93bb6a in ex_execute /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:19730:7
    #24 0xc616d0 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2242:5
    #25 0xc3fe67 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:593:20
    #26 0x91b43c in call_user_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:21608:3
    #27 0x8e50eb in call_func /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6391:11
    #28 0x8fa863 in get_func_tv /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:6148:11
    #29 0x8f33ee in ex_call /home/travis/build/neovim/neovim/build/../src/nvim/eval.c:2743:9
    #30 0xc616d0 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2242:5
    #31 0xc3fe67 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:593:20
    #32 0xc25cb0 in do_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:3224:3
    #33 0xc21ae4 in cmd_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2969:14
    #34 0xc222a0 in ex_source /home/travis/build/neovim/neovim/build/../src/nvim/ex_cmds2.c:2950:3
    #35 0xc616d0 in do_one_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:2242:5
    #36 0xc3fe67 in do_cmdline /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:593:20
    #37 0xc46025 in do_cmdline_cmd /home/travis/build/neovim/neovim/build/../src/nvim/ex_docmd.c:269:10
    #38 0xfd180b in exe_commands /home/travis/build/neovim/neovim/build/../src/nvim/main.c:1699:5
    #39 0xfbf0b1 in main /home/travis/build/neovim/neovim/build/../src/nvim/main.c:548:5
    #40 0x7f2b2c21c82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #41 0x450518 in _start (/home/travis/build/neovim/neovim/build/bin/nvim+0x450518)

0x000001e1b221 is located 63 bytes to the left of global variable '<string literal>' defined in '../src/nvim/eval.c:779:63' (0x1e1b260) of size 2
  '<string literal>' is ascii string '.'
0x000001e1b221 is located 0 bytes to the right of global variable '<string literal>' defined in '../src/nvim/eval.c:777:32' (0x1e1b220) of size 1
  '<string literal>' is ascii string ''
SUMMARY: AddressSanitizer: global-buffer-overflow (/home/travis/build/neovim/neovim/build/bin/nvim+0x4f22e4) in __asan_memcpy
Shadow bytes around the buggy address:
  0x0000803bb5f0: 00 00 00 00 00 00 00 00 00 00 00 00 06 f9 f9 f9
  0x0000803bb600: f9 f9 f9 f9 00 00 00 00 00 00 00 00 00 00 00 00
  0x0000803bb610: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0000803bb620: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0000803bb630: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0000803bb640: 00 00 00 00[01]f9 f9 f9 f9 f9 f9 f9 02 f9 f9 f9
  0x0000803bb650: f9 f9 f9 f9 02 f9 f9 f9 f9 f9 f9 f9 00 00 00 00
  0x0000803bb660: 00 00 00 00 02 f9 f9 f9 f9 f9 f9 f9 03 f9 f9 f9
  0x0000803bb670: f9 f9 f9 f9 00 00 00 00 00 00 00 00 00 00 00 00
  0x0000803bb680: 00 07 f9 f9 f9 f9 f9 f9 00 00 00 00 00 00 00 00
  0x0000803bb690: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
```

janlazo commented:
I attempted these patches in a previous PR but I can't find it. ASAN error is same as before but we have the older patches for `f_spellbadword`. There's no newer 8.1 patches that specifically mention this error.

janlazo commented on src/nvim/testdir/test_spell.vim:
```diff
@@ -68,6 +68,47 @@ func Test_z_equal_on_invalid_utf8_word()
+  set spell
+
+  call assert_equal(['bycycle', 'bad'],  spellbadword('My bycycle.'))
+  call assert_equal(['another', 'caps'], spellbadword('A sentence. another sentence'))
```
Without vim-patch:8.1.0199, this fails.

janlazo replied:
`printf('%ld vs %ld\n', strnlen(word), len)` output `0 vs 6`.

justinmk commented:
7c956dcbe8017dd68bf4221893bf554302b8fd90 changed the return type of `spell_check` to `size_t`. That was a mistake, it should be `ptrdiff_t`.

(Normally that author was very careful about his changes, so I don't think this is a widespread issue.)

justinmk commented:
Maybe this will give some insight:

```diff
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
index 0961e5fd921f..dc355bb7961f 100644
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -15675,6 +15675,7 @@ static void f_spellbadword(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     }
   }
 
+  assert(strlen(word) == len);
   assert(len <= INT_MAX);
   tv_list_alloc_ret(rettv, 2);
   tv_list_append_string(rettv->vval.v_list, word, len);
```

justinmk commented:
No that probably just segfaults at the `strlen()` call. Seems like the `str` produced by `spell_check` is invalid... but difficult to see how any value of `capcol` could cause that.

Perhaps we can work around this by always calculating the string length:

```diff
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
index 0961e5fd921f..da6938106f6f 100644
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -15677,7 +15677,7 @@ static void f_spellbadword(typval_T *argvars, typval_T *rettv, FunPtr fptr)
 
   assert(len <= INT_MAX);
   tv_list_alloc_ret(rettv, 2);
-  tv_list_append_string(rettv->vval.v_list, word, len);
+  tv_list_append_string(rettv->vval.v_list, word, 0);
   tv_list_append_string(rettv->vval.v_list,
                         (attr == HLF_SPB ? "bad"
                          : attr == HLF_SPR ? "rare"
```

If that works, let's call it good. `f_spellbadword` is not a bottleneck of any sort :) The issue is likely somewhere in the pointer-arithmetic hell, and it _very possibly_ is a bug in Vim itself (the ASAN step in Vim's CI does not have all checks enabled, I don't remember which ones though).

janlazo commented:
Should `capcol` have type `ptrdiff_t` because of pointer arithmetic?

justinmk commented:
==29606==ERROR: AddressSanitizer: global-buffer-overflow

Assuming that 29606 is the PID, we could use `ELOG()` to print values before the segfault, then compare the PID in the log to the PID from ASAN.

```diff
diff --git a/src/nvim/eval.c b/src/nvim/eval.c
index 0961e5fd921f..c353a078adaa 100644
--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -15675,6 +15675,7 @@ static void f_spellbadword(typval_T *argvars, typval_T *rettv, FunPtr fptr)
     }
   }
 
+  ELOG("len=%td strlen=%zu word=%s", len, strlen(word), word);
   assert(len <= INT_MAX);
   tv_list_alloc_ret(rettv, 2);
   tv_list_append_string(rettv->vval.v_list, word, len);
```

janlazo commented:
I don't see `ELOG` for the oldtests.

justinmk commented:
> I don't see ELOG log output for the oldtests.

Ah, right, that's missing.  Looks like this code is not exercised in any Lua tests...

Fastest thing at this point would be to step through the code in gdb.

janlazo commented:
Anything else regressed in https://github.com/neovim/neovim/commit/7c956dcbe8017dd68bf4221893bf554302b8fd90?

```
int find_ident_under_cursor(...) --> size_t find_ident_under_cursor(...)
int find_ident_at_pos(...) --> size_t find_ident_at_pos(...)
```
What about these?

janlazo commented on src/nvim/normal.c:
```diff
@@ -4500,7 +4500,7 @@ static void nv_zet(cmdarg_T *cap)
        * used below. */
       emsg_off++;
-      len = spell_move_to(curwin, FORWARD, true, true, NULL);
+      len = (size_t)spell_move_to(curwin, FORWARD, true, true, NULL);
```
I guess this is what fails the test but changing len to `ptrdiff_t` is hard.

justinmk replied:
would it help to disable `-Wconversion` for now?

blueyed commented:
I can see this locally using both the current commit, or the previous (without the revert).
The problem is:
```
#9  0x0000559c057a9c68 in f_spellbadword (argvars=0x7ffda174dba0, rettv=0x7ffda17500d0, fptr=0x0) at ../src/nvim/eval.c:16743
16743     tv_list_append_string(rettv->vval.v_list, word, len);
(gdb) p word
$4 = 0x559c06a1d5c0 <str> ""
(gdb) p len
$5 = 6
```

I.e. "len" is 6, but "word" only "".

Minimal: `build/bin/nvim -u NONE -c 'set spell' -c 'call spellbadword("good")'`,

This appears to fix it:
```diff
diff --git i/src/nvim/eval.c w/src/nvim/eval.c
index 4e2a2c048..bf81993ef 100644
--- i/src/nvim/eval.c
+++ w/src/nvim/eval.c
@@ -16734,6 +16734,7 @@ static void f_spellbadword(typval_T *argvars, typval_T *rettv, FunPtr fptr)
         }
         str += len;
         capcol -= len;
+        len = 0;
       }
     }
   }
```
i.e. set length back to 0 when `word` is not being set.

Don't know why that is not a problem in Vim though.. (with their ASAN build)

blueyed commented:
Vim's `list_append_string` uses `vim_strnsave`, which is ok with `str=""` and `len=4`.
Nvim's `tv_list_append_string` uses `xmemdupz(str, (size_t)len)`, which is not ok with that then.

I think the `len = 0` fix is legit (will create a Vim PR), but maybe we can improve error handling / assertions in that case?

blueyed committed:
f_spellbadword: set len=0 for non-found word

`len` is used with `list_append_string` later, and should reflect the
length of `word` (i.e. 0 when not setting word / breaking above).

Ref: neovim/neovim#9782 (comment)

janlazo commented:
@blueyed Did you check if any of the functions, refactored from int to size_t (or ssize_t) in https://github.com/neovim/neovim/commit/7c956dcbe8017dd68bf4221893bf554302b8fd90, return a negative value on Vim?

blueyed commented:
> Did you check if any of the functions, refactored from int to size_t (or ssize_t) in 7c956dc, return a negative value on Vim?

No.
